Delete lot ACAS-49

brianbolt opened the pull request:
## Description
 - change default `client.cmpdreg.serverSettings.projectRestrictions` to `true`
 - Add DeleteLotController to CmpdReg
 - Add `Download SDF` button to CmpdReg lot details page to download SDF
 - Add route to delete lots which checks lot dependencies and user acls
 - 

## Related Issue
ACAS-49

## How Has This Been Tested?
ACAS client tests. Plus testing via GUI by using breakpoints in acasclient tests to grab user creds with specific roles and test UI functionality.

 - get acas user from debug spot

https://user-images.githubusercontent.com/868119/178833387-af84aa8e-d1f3-4671-9f17-a26c9b1bee43.mov

 - acas user can't click register and can't see restricted compound `CMPD-0000002-002`

https://user-images.githubusercontent.com/868119/178833632-21ad17b7-5a1c-4116-9729-706fd5242239.mov

 - cmpdreg user with with edit access to the lot (belong to project and own the lot) but can't delete because they aren't an acas user with access to the experiment

https://user-images.githubusercontent.com/868119/178835567-590e0ddc-b6eb-40d9-b263-19d77c9c5088.mov

 - bob is god
  - Bob can view it all

https://user-images.githubusercontent.com/868119/178837905-9fb1414b-394b-43f3-9144-a2995ef302a2.mov

  - Bob can delete it all

https://user-images.githubusercontent.com/868119/178837961-e621c6c6-9aa1-42aa-888e-5f135feef14d.mov

brianbolt commented on conf/config.properties.example:
```diff
@@ -725,7 +725,7 @@ client.cmpdreg.serverSettings.maxSearchTime=180000
 client.cmpdreg.serverSettings.unitTestDB=false
 client.cmpdreg.serverSettings.initalDBLoad=false
-client.cmpdreg.serverSettings.projectRestrictions=false
+client.cmpdreg.serverSettings.projectRestrictions=true
```
We should have these enabled by default. I checked every prod server and they are all set to true.

dalejerikson reviewed (commented):
Super readable and thorough. I'm enjoying reading through these PRs 🥇

bffrost commented on modules/ServerAPI/src/server/routes/ExperimentServiceRoutes.coffee:
```diff
@@ -785,41 +785,171 @@ exports.createAndUpdateExptExptItxs = (req, resp) ->
+			return false
+	return true
+
+exports.getEntityValue = (entity, stateType, stateKind, valueType, valueKind) ->
```
This feels more like a global utility, no?

brianbolt replied:
Yeah, but waiting to move it there until it gets reused.

bffrost reviewed (commented):
This is a large and complex PR. I've done my best to review it, but it was hard to do it all at once. It would've been nice to maybe split this up into a PR for the GUI and basic deleting workflow, a PR for the ACL enhancements, and maybe a 3rd PR for some of the refactoring you've done.

Left some comments on things I had questions about, but overall this looks good to me. Given the size and complexity, we'll want to merge it fairly soon to give QA enough time to explore this new feature and all the ACL complexities.

brianbolt commented on modules/CmpdReg/src/client/src/MetaLot.js:
```diff
@@ -312,6 +319,39 @@ $(function() {
+		    });
+	    },
+
+	    delete: function () {
```
It didn't get called, this was my initial implmentation which I was going to directly in this controller but changed course.  I just removed it. Thanks.

brianbolt commented on modules/CmpdReg/src/client/DeleteLot.coffee:
```diff
@@ -0,0 +1,229 @@
+
+
+	initialize: ->
+		_.bindAll(@, 'handleCancelButtonClicked', 'handleDeleteButtonClicked', 'checkDependencies', 'dependencyCheckReturn', 'dependencyCheckError', 'deleteLotError', 'deleteLotReturn', 'downloadLot', 'handleBackToCregButtonClicked');
```
Mentioned this in a comment above but yes you are right:

> This is necessary because we are using older backbone which didn't bind @ to the view attributes.

brianbolt commented on modules/CmpdReg/src/server/routes/CmpdRegRoutes.coffee:
```diff
@@ -223,7 +227,133 @@ checkStatus = (response) ->
+			# This way it'know there are experiments linked that aren't readable
+			if !acls.getRead()
+				console.log "Experiment #{experiment.experiment.codeName} is not readable by user #{user.username}"
+				# Get the experiment from the linkedExperiments by code
```
Thanks added a bunch of comments around this

brianbolt commented:
I've made one more code change beyond the PR comments which turns of linkedLots from being fetched.  Linked lots aren't a requirement for deleting a lot so it's purely informational when checking for acls to delete.  Linked lots are returned by default in both the internal function and by the route but because it can be unnecessary when checking delete acls, I've added the parameter to turn it off.

Re-requesting review from @bffrost to look at this and because I've fixed PR comments he requested.
